Qiskit/qiskit-dynamics
# Add RK4 fixed step solver and parallelized JAX fixed step solvers

### Summary

The goal of this PR is to:
- Add jax and numpy-based RK4 fixed step solver
- Add a new "fixed step LMDE solver template" for JAX that uses parallel operations, i.e. all time-steps are integrated simultaneously in parallel using `vmap`.
    - This is specific to LMDEs as it assumes that structure for the parallelization to work.
    - This method works by simulating the full propagator over each time-step in parallel, then multiplying them all together (also in parallel, using `jax.lax.associative_scan`). A consequence of this is that, if the user wants to simulate only a single state, this approach will actually be more expensive than serial computations (in terms of linear algebraic operations), and may not be worth the parallelization. However, if a user wants to simulate the full propagator itself, then no additional computational cost is created.
- Using the above, add jax parallel versions of both RK4 and expm solvers.


### Details and comments

Currently implemented the jax parallel template, the jax parallel expm solver, and have added tests for the parallel expm solver derived directly from the already existing tests for expm solvers through inheritance. The template code currently assumes that `y0` is square, so it needs to be expanded, but the tests are passing so this is a good start!

To do:
- ~Generalize the parallel template to accept non-square `y0`.~
    - ~Test!~
- ~Add RK4 solvers~
    - ~Test! (Can probably abstract out content of expm tests.)~
- Add proper interface access to new solvers through `solve_lmde` and `solve_ode`.
    - Add some higher level tests for these.
    - Update documentation.
- Check if update necessary in module docs.
